### PR TITLE
ENH: Added asymptotic expansion for invgamma entropy (#18093)

### PR DESCRIPTION
**Pull request by MatteoRaso**
#### Reference issue

Part of #18093 
#### What does this implement/fix?

Adds asymptotic expansion for invgamma entropy

### CODE REVIEW

File: scipy/stats/_continuous_distns.py
Comment (melissawm) [inline scipy/stats/_continuous_distns.py:4357]: I am not familiar with these formulas, but nitpicking I would suggest calling these functions something like "regular_formula" and "asymptotic formula" as you see in https://github.com/scipy/scipy/pull/18095/files, for example. Also can you clarify the source of the asymptotic formula for those like me? 🙂 Thanks!

**Review by melissawm — CHANGES_REQUESTED** [review]

**Review by dschmitz89 — CHANGES_REQUESTED** [review]
Thanks for the PR! Looking at the last PR in this area (#17930) for acceptance, we need two changes:
* more terms in the expansion
* an error analysis of expansion versus the regular formula, so that we can set the cutoff as low as possible

**Comment by MatteoRaso** [comment]
> Thanks for the PR! Looking at the last PR in this area (#17930) for acceptance, we need two changes:
> 
>     * more terms in the expansion
> 
>     * an error analysis of expansion versus the regular formula, so that we can set the cutoff as low as possible

Okay, I've added an additional term, which decreased the error for 1e8 and above to less than 1e-10. As for the error analysis, it seems that the cutoff to start using the asymptotic expansion was a bit lower than I expected, at ~5e7 instead of 1e8.
Figure_1

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py]: I've changed the function names. As for the source of the formula, I took the expansions listed in the linked issue and plugged them in.

**Comment by dschmitz89** [comment]
Thanks for the PR! Looking at the last PRs in this area, could you add

> > Thanks for the PR! Looking at the last PR in this area (#17930) for acceptance, we need two changes:
> > ```
> > * more terms in the expansion
> > 
> > * an error analysis of expansion versus the regular formula, so that we can set the cutoff as low as possible
> > ```
> 
> Okay, I've added an additional term, which decreased the error for 1e8 and above to less than 1e-10. As for the error analysis, it seems that the cutoff to start using the asymptotic expansion was a bit lower than I expected, at ~5e7 instead of 1e8. Figure_1

Thanks for the updates: regarding an error analysis I was referring to a comparison of the relative error of both formulas. See [here](https://github.com/scipy/scipy/pull/17930#discussion_r1122526573) or [here, lower plot](https://github.com/scipy/scipy/pull/18095#issue-1609943166). That will allow us to decide for the best cutoff.

**Comment by MatteoRaso** [comment]
> Thanks for the PR! Looking at the last PRs in this area, could you add
> 
> > > Thanks for the PR! Looking at the last PR in this area (#17930) for acceptance, we need two changes:
> > > ```
> > > * more terms in the expansion
> > > 
> > > * an error analysis of expansion versus the regular formula, so that we can set the cutoff as low as possible
> > > ```
> > 
> > 
> > Okay, I've added an additional term, which decreased the error for 1e8 and above to less than 1e-10. As for the error analysis, it seems that the cutoff to start using the asymptotic expansion was a bit lower than I expected, at ~5e7 instead of 1e8. Figure_1
> 
> Thanks for the updates: regarding an error analysis I was referring to a comparison of the relative error of both formulas. See [here](https://github.com/scipy/scipy/pull/17930#discussion_r1122526573) or [here, lower plot](https://github.com/scipy/scipy/pull/18095#issue-1609943166). That will allow us to decide for the best cutoff.

Here you go.
Figure_1

**Comment by MatteoRaso** [comment]
More zoomed in around 1e7:
Figure_1
Looks like 1e7 is a better cutoff.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:4351]:
```suggestion
            h = a - (a+1.0) * sc.psi(a) + sc.gammaln(a)
```

This would usually have been caught by the linter, no idea what is going on there at the moment.

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:2937]:
```suggestion
        # def invgamma_entropy(a):
        #     a = mp.mpf(a)
        #     h = a + mp.loggamma(a) - (mp.one + a) * mp.digamma(a)
```
```suggestion
        # def invgamma_entropy(a):
        #     h = a + mp.loggamma(a) - (mp.one + a) * mp.digamma(a)
```

Does this change the result?

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:2930]: Two very high values should be fine, we do not need five. What about we add also tests for small `a` if we are anyway adding tests for the entropy here?

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:4358]:
```suggestion
        h = _lazywhere(a >= 8e6, (a), f=asymptotic, f2=regular)
```

Juding from the [plot](https://github.com/scipy/scipy/pull/18122#issuecomment-1464501670), this seems appropriate to me as the error of the asymptotic expansion is lower already.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [reply scipy/stats/_continuous_distns.py]: The employed asymptotic expansions should go in comments or at least the PR description. We did this also for the other distributions (see [here](https://github.com/dschmitz89/scipy/blob/76dca342b63357d0ebe472796d1f0feff861ad66/scipy/stats/_continuous_distns.py#L2735) for example). Otherwise it will be hard to maintain in future.

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py:2937]: Doesn't seem to, no.

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:4358]: No, zooming into the graph, the main formula is definitely more accurate overall than the asymptotic expansion at 8e6.
Figure_1

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py]: Ok, I've added the formulas as comments.

File: scipy/stats/tests/test_distributions.py
Comment (MatteoRaso) [reply scipy/stats/tests/test_distributions.py]: Ok, I've removed the middle 3 values. As for the small values, like I said in the PR for gengamma entropy, I'd only do that in a different PR because this one is only meant for large values.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [reply scipy/stats/_continuous_distns.py:4360]: Thansk for the detailed plot. Can we agree on the cutoff `8.5e6` instead? From there, the expansion is defintely better.

Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:4360]: >Can we agree on the cutoff 8.5e6 instead? From there, the expansion is defintely better.

The area of the main error that's underneath the extra term line seems to be significantly larger than the area above it. Maybe it's safe to cutoff at 9e6, but 1e7 seems best to me.

Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:4360]: Another zoomed in plot
Figure_1

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:2934]: Let's still add it so that future contributors use mpmath the proper way.

File: scipy/stats/_continuous_distns.py
Comment (dschmitz89) [inline scipy/stats/_continuous_distns.py:4352]:
```suggestion
            h = a - (a + 1.0) * sc.psi(a) + sc.gammaln(a)
            return h
```

As in other places, a problem with the current linting infrastructure.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

File: scipy/stats/tests/test_distributions.py
Comment (dschmitz89) [inline scipy/stats/tests/test_distributions.py:2927]:
```suggestion
                             [(100000000.0, -26.21208257605721),
                              (1e+100, -343.9688254159022)])
```

Sorry for the noise, the linter is dearly missed :(.

After this, I will approve.

**Review by dschmitz89 — CHANGES_REQUESTED** [review]

**Review by dschmitz89 — APPROVED** [review]

**Comment by mdhaber** [comment]
Similar to gh-18094, please use a program to derive these things and post the results.

```python3
import sympy
from sympy import log, pi, Rational, simplify, expand

x, a = sympy.symbols('x, a')

def digamma(x):
    return log(x) - 1/(2*x) - 1/(12*x**2) + 1/(120*x**4)

def gammaln(x):
    return x*log(x) - x - log(x)/2 + log(2*pi)/2 + 1/(12*x) - 1/(360*x**3)

old = a - (a + 1) * digamma(a) + gammaln(a)

print(expand(simplify(old)))
# -3*log(a)/2 + log(2)/2 + 1/2 + log(pi)/2 + 2/(3*a) + 1/(12*a**2) - 1/(90*a**3) - 1/(120*a**4)
```

With the formula corrected, we get error like:
image

So a good cutoff is `2e2`.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4357]: The powers of `a` can overflow, but that means that the result is negligible. This could be fixed by evaluating them in logspace first, or:
```suggestion
            with np.errstate(over='ignore'):
                h = ((1 - 3*np.log(a) + np.log(2) + np.log(np.pi))/2
                     + 2/(3*a) + 1/(12*a**2) - 1/(90*a**3) - 1/(120*a**4))
            return h
```

File: scipy/stats/tests/test_distributions.py
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2937]:
```suggestion
        assert_allclose(stats.invgamma.entropy(a), ref, rtol=1e-15)
```

**Review by mdhaber — CHANGES_REQUESTED** [review]

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4359]:
```suggestion
```

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4349]: Need to check this out.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4362]:
```suggestion
        h = _lazywhere(a >= 2e2, (a,), f=asymptotic, f2=regular)
```

**Review by mdhaber — COMMENTED** [review]
The second argument to `_lazywhere` should be a tuple. 

Converting to an array isn't necessary for the calculations. Turns out `a` is always a scalar because `entropy` applies `np.vectorize` to `_entropy`. So even the `_lazywhere` is unnecessary.

**Comment by mdhaber** [comment]
Those were all supposed to go in a batch.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4361]:
```suggestion
        return h

```
PEP8 but let's NOT merge this one if CI doesn't complain. Not worth another run. Please fix it in another PR.

**Review by mdhaber — APPROVED** [review]

File: scipy/stats/_continuous_distns.py
Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:4361]: CI won't complain, the linter's broken (unless it's been fixed very recently.)

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4361]: If it's broken, that must be very recent. IIUC, we switched linters about a month ago and it not currently checking all the rules we used to.

Comment (MatteoRaso) [reply scipy/stats/_continuous_distns.py:4361]: Then let's not worry about the PR until the linter is fully up to snuff. Also, a PR just for a single pep8 violation seems a bit wasteful. It feels like it's be better to try and catch all the violations in one big PR once the linter can catch them all.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:4361]: I mean in any one of the other PRs you have open.

File: scipy/stats/_continuous_distns.py
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:4357]: In that same PR, please add decimal points to these integer powers to avoid `ValueError: Integers to negative integer powers are not allowed.`

**Pull request merged**
